sled-hardware/src/illumos/partitions.rs

```diff
@@ -154,13 +231,124 @@ fn internal_ensure_partition_layout<GPT: gpt::LibEfiGpt>(
     }
 }
 
+fn ensure_size_and_formatting(
```

Comment:
Is there any problem running this over after failure on a reboot or restart of sled-agent?

Reply:
So while testing I found that if a disk is left in the unattached state then sled-agent will ignore it.  That's due to [these lines](https://github.com/oxidecomputer/omicron/blob/8697f39aac3323ffda3f33f4b5aa44fd7ad857e3/sled-hardware/src/illumos/mod.rs#L435-L437).  So, if we were to detach blkdev from the disk and then take a sled-agent reboot we wouldn't be notified of that disk again I believe.  

Now if we were in the process of formatting a disk and the box panicked or rebooted for some reason I don't know what would happen.  Perhaps @rmustacc could shed some light on that failure scenario.

Regardless I don't know what the best course of action is for the first failure mode, maybe we need an enhancement in how we detect disks from devinfo?

Reply:
> Now if we were in the process of formatting a disk and the box panicked or rebooted for some reason I don't know what would happen. Perhaps @rmustacc could shed some light on that failure scenario.

There is nothing persistent about a blkdev attach/detach, so the kernel will attempt to if it can. Because we don't support namespace management and aren't doing anything here, this should likely be okay, though I can't comment on what happens if a device fails in the middle of say a format nvm command.

> Regardless I don't know what the best course of action is for the first failure mode, maybe we need an enhancement in how we detect disks from devinfo?

I would suggest we file a follow up bug that moves this to looking for the NVMe node and not the blkdev node. This is for a few reasons:

* An NVMe device may not have any blkdev instances attached as mentioned to start this.
* An NVMe device may have the wrong number of namespaces created if it came from another system. This would show up as two distinct block devices or possibly no block devices, which would likely be confusing.